### DISCUSSION TITLE
When data including missing values are submitted to the mcar() function of the mice package, the message "No need for mice. This data set is completely observed." is sometimes displayed.

### DISCUSSION THREAD

**Discussion by indenkun** [amices/mice#605] [Q&A]
When I submit data with missing values to the mcar() function of the mice package, the message "No need for mice. This data set is completely observed." is sometimes displayed.

For example, if you submit missvals data from the mvnmle package, you will get the following message.

``` r
library(mice)
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(mvnmle)
data("missvals")
mcar(missvals)
#> 
#>  iter imp variable
#>   1   1  x1  x2  x4
#>   1   2  x1  x2  x4
#>   1   3  x1  x2  x4
#>   1   4  x1  x2  x4
#>   1   5  x1  x2  x4
#>   2   1  x1  x2  x4
#>   2   2  x1  x2  x4
#>   2   3  x1  x2  x4
#>   2   4  x1  x2  x4
#>   2   5  x1  x2  x4
#>   3   1  x1  x2  x4
#>   3   2  x1  x2  x4
#>   3   3  x1  x2  x4
#>   3   4  x1  x2  x4
#>   3   5  x1  x2  x4
#>   4   1  x1  x2  x4
#>   4   2  x1  x2  x4
#>   4   3  x1  x2  x4
#>   4   4  x1  x2  x4
#>   4   5  x1  x2  x4
#>   5   1  x1  x2  x4
#>   5   2  x1  x2  x4
#>   5   3  x1  x2  x4
#>   5   4  x1  x2  x4
#>   5   5  x1  x2  x4
#>  /\     /\
#> {  `---'  }
#> {  O   O  }
#> ==>  V <==  No need for mice. This data set is completely observed.
#>  \  \|/  /
#>   `-----'
#> Warning in base::rbind(...): number of columns of result is not a multiple of
#> vector length (arg 2)
#> Error in pats[-nrow(pats), colnames(missings)]: no 'dimnames' attribute for array
```

<sup>Created on 2023-11-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

~~This "No need for mice. This data set is completely observed." seems to be in the md.pattern() function, but when I submit missvals from the mvnmle package to md.pattern(), this message does not appear.~~

~~Is there a way around this message?~~

After reading the source code, I found that this message problem can be resolved by setting min_n to less than the default value of 6.

``` r
library(mice)
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
library(mvnmle)
data("missvals")
mcar(missvals, min_n = 3)
#> 
#>  iter imp variable
#>   1   1  x1  x2  x4
#>   1   2  x1  x2  x4
#>   1   3  x1  x2  x4
#>   1   4  x1  x2  x4
#>   1   5  x1  x2  x4
#>   2   1  x1  x2  x4
#>   2   2  x1  x2  x4
#>   2   3  x1  x2  x4
#>   2   4  x1  x2  x4
#>   2   5  x1  x2  x4
#>   3   1  x1  x2  x4
#>   3   2  x1  x2  x4
#>   3   3  x1  x2  x4
#>   3   4  x1  x2  x4
#>   3   5  x1  x2  x4
#>   4   1  x1  x2  x4
#>   4   2  x1  x2  x4
#>   4   3  x1  x2  x4
#>   4   4  x1  x2  x4
#>   4   5  x1  x2  x4
#>   5   1  x1  x2  x4
#>   5   2  x1  x2  x4
#>   5   3  x1  x2  x4
#>   5   4  x1  x2  x4
#>   5   5  x1  x2  x4
#> 
#> Missing data patterns: 2 used, removed.
#> Cases used: 10 
#> 
#> Hawkins' test: median chi^2 (4) = 3.017096, median p = 0.5549686. Some p-values for Hawkins' test were significant; please inspect their values, e.g., using `plot(x)`
#> 
#> Anderson-Darling rank test: median T = 1.330522, median p = 0.1951404. Some p-values for the Anderson-Darling test were significant; please inspect their values, e.g., using `plot(x)`
#> 
#> Interpretation of results:
#>  Hawkins' test is not significant; there is no evidence to reject the assumptions of multivariate normality and MCAR.
#>  Anderson-Darling test is not significant. There is thus evidence against multivariate normality, but not against MCAR.
```

<sup>Created on 2023-11-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

How should I go about setting a statistically reasonable value for this min_n? Can it be reduced until the message is resolved and a solution is obtained, as in this case?

**Comment by stefvanbuuren** [comment]
@cjvanlissa Could you take a look at this?

**Comment by indenkun** [comment]
By the way, for the default value of 6 for min_n, see, 
Jamshidian, M., & Jalal, S. (2010). Tests of homoscedasticity, normality, and missing completely at random for incomplete multivariate data. Psychometrika, 75(4), 649- 674. doi:10.1007/s11336-010-9175-3 in

> Moreover, we have removed missing data patterns that included 6 or less cases.

I thought it was derived from this statement, is this perception incorrect? (I couldn't understand why it was 6, though.)

By the way, if I make min_n smaller (trying to get the MCAR test done with all the data), I will get the following error.
``` r
library(mice)
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
mcar(airquality, min_n = 1)
#> 
#>  iter imp variable
#>   1   1  Ozone  Solar.R
#>   1   2  Ozone  Solar.R
#>   1   3  Ozone  Solar.R
#>   1   4  Ozone  Solar.R
#>   1   5  Ozone  Solar.R
#>   2   1  Ozone  Solar.R
#>   2   2  Ozone  Solar.R
#>   2   3  Ozone  Solar.R
#>   2   4  Ozone  Solar.R
#>   2   5  Ozone  Solar.R
#>   3   1  Ozone  Solar.R
#>   3   2  Ozone  Solar.R
#>   3   3  Ozone  Solar.R
#>   3   4  Ozone  Solar.R
#>   3   5  Ozone  Solar.R
#>   4   1  Ozone  Solar.R
#>   4   2  Ozone  Solar.R
#>   4   3  Ozone  Solar.R
#>   4   4  Ozone  Solar.R
#>   4   5  Ozone  Solar.R
#>   5   1  Ozone  Solar.R
#>   5   2  Ozone  Solar.R
#>   5   3  Ozone  Solar.R
#>   5   4  Ozone  Solar.R
#>   5   5  Ozone  Solar.R
#> Error in mcar.data.frame(airquality, min_n = 1): object 'idpats' not found
```

<sup>Created on 2023-12-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I think this is due to the following part of the mcar() code;

```r
remove_pats <- as.numeric(rownames(pats))[-nrow(pats)] <= min_n
  if (any(remove_pats)) {
  # omission
    idpats <- do.call(paste, as.data.frame(pats == 0))
  # omission
  }
  # omission
  idpats <- idpats[!remove_pats]
```

I thought this was because idpats assumes that some data will be deleted, and if it is not deleted, it is trying to define idpats using idpats that do not exist yet.

Is this due to some statistical reason that this process should always delete more than 1 pattern of data? However, this seems somewhat inconsistent with the fact that min_n accepts 1. (Incidentally, I checked the behavior of MissMech::TestMCARNormality(), which is described in the documentation as the source of the re-implementation, and this one was able to perform the mcar test using all data. I could not determine which process was more valid.)
